File: testerum-backend/testerum-runner/testerum-runner-cmdline/src/main/assembly/report_templates/common/testerum-model/model/step/def/undefined-step-def.ts
import {StepDef} from "./step-def";
import {Path} from "../../path";
import {StepPhaseEnum} from "./step-phase-enum";
import {StepPattern} from "./pattern/step-pattern";
import {MarshallingUtils} from "../../../json-marshalling/marshalling-utils";

export class UndefinedStepDef implements StepDef {

    constructor(public readonly id: string,
                public readonly path: Path,
                public readonly phase: StepPhaseEnum,
                public readonly stepPattern: StepPattern,
                public readonly description: string | null,
                public readonly tags: Array<string>) { }

    static parse(input: Object): UndefinedStepDef {
        const id = input["id"];
        const path = Path.parse(input["path"]);
        const phase = MarshallingUtils.parseEnum(input["phase"], StepPhaseEnum);
        const stepPattern = StepPattern.parse(input["stepPattern"]);
        const description = input["description"];
        const tags = MarshallingUtils.parseListOfStrings(input["tags"]);

        return new UndefinedStepDef(id, path, phase, stepPattern, description, tags);
    }

}

File: model/src/main/java/com/testerum/model/step/UndefinedStepDef.kt
package com.testerum.model.step

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.testerum.model.enums.StepPhaseEnum
import com.testerum.model.infrastructure.path.Path
import com.testerum.model.text.StepPattern
import com.testerum.model.util.StepHashUtil
import com.testerum.model.warning.Warning

data class UndefinedStepDef @JsonCreator constructor(@JsonProperty("phase") override val phase: StepPhaseEnum,
                                                     @JsonProperty("stepPattern") override val stepPattern: StepPattern): StepDef {
    override val id: String
        get() = StepHashUtil.calculateStepHash(phase, stepPattern)

    override val path: Path
        get() = Path(emptyList(), null, null)

    override val description: String?
        get() = null

    override val tags: List<String>
        get() = emptyList()

    override val warnings: List<Warning>
        get() = emptyList() // undefined steps don't have warnings; "undefined step" is a warning of a step call, not a step definition

    @get:JsonProperty("descendantsHaveWarnings")
    override val descendantsHaveWarnings: Boolean
        get() = false // undefined steps don't have descendants


    override fun toString() = buildString {
        append(phase)
        append(" ")
        append(stepPattern.toDebugString(varPrefix = "<<", varSuffix = ">>"))
    }

}
